tun/tun_illumos.go

```diff
@@ -0,0 +1,519 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2019 Joyent, Inc.
+ */
+// +build illumos,amd64
+
+package tun
+
+import (
+	"fmt"
+
+	"unsafe"
+	"syscall"
+
+	"os"
+//	"os/exec"
+	"golang.org/x/sys/unix"
+)
+
+/*
+ * ILLUMOS SYSTEM CALL HELPERS
+ */
+
+//go:cgo_import_dynamic libc_ioctl ioctl "libc.so"
+//go:cgo_import_dynamic libc_putmsg putmsg "libc.so"
+//go:cgo_import_dynamic libc_getmsg getmsg "libc.so"
+
+//go:linkname f_ioctl libc_ioctl
+//go:linkname f_putmsg libc_putmsg
+//go:linkname f_getmsg libc_getmsg
+
+var (
+	f_ioctl uintptr
+	f_putmsg uintptr
+	f_getmsg uintptr
+)
+
+const (
+	/*
+	 * For use with "/dev/tun":
+	 */
+	TUNNEWPPA = 0x540001
+	TUNSETPPA = 0x540002
+
+	/*
+	 * sys/stropts.h:
+	 */
+	I_STR = 0x5308
+	I_POP = 0x5303
+	I_PUSH = 0x5302
+	I_PLINK = 0x5316
+	I_PUNLINK = 0x5317
+
+	/*
+	 * sys/sockio.h:
+	 */
+	IF_UNITSEL = 0x80047336 /* set unit number */
+
+	SIOCSLIFMUXID = 0x80786984
+	SIOCGLIFMUXID = 0xc0786983
+	SIOCGLIFINDEX = 0xc0786985
+)
+
+func sysvicall6(trap, nargs, a1, a2, a3, a4, a5, a6 uintptr) (
+    r1, r2 uintptr, err syscall.Errno)
+
+func ioctl(fd int, req uint, arg uintptr) (r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_ioctl)), 3,
+	    uintptr(fd), uintptr(req), arg, 0, 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+func putmsg(fd int, ctlptr uintptr, dataptr uintptr, flags int) (
+    r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_putmsg)), 4,
+	    uintptr(fd), ctlptr, dataptr, uintptr(flags), 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+func getmsg(fd int, ctlptr uintptr, dataptr uintptr, flagsp uintptr) (
+    r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_getmsg)), 4,
+	    uintptr(fd), ctlptr, dataptr, flagsp, 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+
+/*
+ * WIREGUARD/TUN INTERFACE
+ */
+
+type NativeTun struct {
```

Comment:
It would be good Go style to add docstrings for exported names, but the rest of this package doesn't do that either. If you're feeling zealous, adding them would make this codebase more idiomatic Go. e.g.

`// NativeTun is the OS-specific implementation of a tun interface.`

Not required though, since as I said the rest of this package doesn't follow that convention.

Reply:
Please don't do that here, but if you (@danderson) want to go through and add documentation to places in a consistent manner, please submit that as a different PR; it'd be welcome.

Reply:
I meant adding docstrings just to the illumos files, rather than bundle in unrelated changes to the other files. Happy to ship documentation for the other bits separately, yeah.

Reply:
Would still love to have mentioned docs at some point.